### ISSUE TITLE
Feature: Mac-specific optimizations for production deployment

### ISSUE THREAD

**Issue by DrDrewCain** [DrDrewCain/MiniMacLLM#13]
## Vision Alignment

For therapeutic use, the model must run **efficiently on consumer Macs** without requiring expensive hardware. Current model works but needs optimization for real-world deployment.

## Current Status

✅ **Working:** Runs on M3 Max (36GB RAM)  
⚠️ **Issues:**
- Requires high-end Mac
- 15-20GB memory during training
- ~2-3GB inference (acceptable but improvable)
- No quantization support
- MPS not fully optimized

## Goal

Enable MiniMacLLM to run smoothly on **entry-level Macs** (M1, 8GB RAM) for therapeutic applications.

### 1. Memory Footprint
**Current:** 15-20GB training, 2-3GB inference  
**Target:** 6-8GB training, <1GB inference

### 2. Inference Speed
**Current:** ~50ms/token  
**Target:** <30ms/token (real-time conversational)

### 3. Hardware Support
**Current:** Requires M3 Max (36GB)  
**Target:** M1/M2/M3 with 8GB+ RAM

### A. Quantization (Priority: High)

**INT8 Quantization:**
```python
# Reduce model size by 75%
# 127M params × 4 bytes (fp32) = 508MB
# 127M params × 1 byte (int8) = 127MB
```

**Benefits:**
- 4x less memory
- Faster inference on Apple Silicon
- Maintains 95%+ quality

**Implementation:**
- [ ] Add quantization utilities
- [ ] Benchmark quality vs size trade-offs
- [ ] Create quantized model loading
- [ ] Document memory savings

**INT4 Quantization (Future):**
- 8x less memory
- Useful for very constrained environments

### B. MPS-Specific Optimizations

**Current Issues:**
- Generic PyTorch MPS backend
- No Apple Neural Engine utilization
- Memory allocation not optimal

**Optimizations:**
```python
# Use MPS-optimized operations
# Pre-allocate memory pools
# Batch operations efficiently
# Use Metal Performance Shaders where possible
```

**Implementation:**
- [ ] Profile MPS kernel usage
- [ ] Replace slow ops with MPS-optimized versions
- [ ] Memory pooling for repeated operations
- [ ] Benchmark vs CPU/CUDA

### C. Model Architecture Adjustments

**Smaller Variants:**
```yaml
# Mini (for 8GB Macs)
params: 50M
d_model: 384
layers: 6

# Small (for 16GB Macs)
params: 127M  # Current
d_model: 768
layers: 12

# Medium (for 32GB+ Macs)
params: 350M
d_model: 1024
layers: 16
```

**Implementation:**
- [ ] Create configs for different Mac tiers
- [ ] Benchmark quality across sizes
- [ ] Document hardware requirements

### D. Efficient Inference

**KV Cache Optimization:**
```python
# Current: Store full KV cache
# Optimized: Sliding window + compression
```

**Flash Attention:**
- Integrate Flash Attention 2/3
- Memory-efficient attention computation
- 2-4x speedup on longer contexts

**Speculative Decoding:**
- Use tiny draft model + verification
- 2-3x faster generation
- Useful for therapeutic conversations

### E. Memory Management

**Gradient Checkpointing (Training):**
- Already implemented
- Can be tuned further

**Mixed Precision (BF16):**
```python
# Use bfloat16 instead of float32
# 2x memory savings
# Native Apple Silicon support
```

**Lazy Loading:**
- Load model weights on-demand
- Reduces startup memory

## Benchmarking Framework

Create standardized benchmarks:

```python
# scripts/benchmark_mac.py
- Measure: memory, speed, quality
- Test on: M1, M2, M3 (8GB, 16GB, 32GB configs)
- Report: JSON with all metrics
```

**Metrics:**
- Peak memory (training/inference)
- Tokens/second
- Cold start time
- Model load time
- Generation quality (perplexity)

### Phase 1: Measurement (Week 1)
- [ ] Create benchmark suite
- [ ] Profile current performance
- [ ] Identify bottlenecks

### Phase 2: Quick Wins (Week 2)
- [ ] INT8 quantization
- [ ] Mixed precision (BF16)
- [ ] Memory pool optimization

### Phase 3: Architecture (Week 3-4)
- [ ] Create mini/small/medium configs
- [ ] Flash Attention integration
- [ ] KV cache optimization

### Phase 4: Validation (Week 5)
- [ ] Benchmark all variants
- [ ] Quality vs performance trade-offs
- [ ] Documentation updates

## Success Criteria

**Hard Requirements:**
- [ ] Runs on M1 with 8GB RAM (mini model)
- [ ] <1GB inference memory (quantized)
- [ ] <30ms/token generation
- [ ] <95% quality degradation from base

**Stretch Goals:**
- [ ] Apple Neural Engine utilization
- [ ] <500MB inference memory (int4)
- [ ] <20ms/token generation
- [ ] Multi-modal support preparation

## Hardware Test Matrix

| Mac Model | RAM | Model Size | Expected Performance |
|-----------|-----|------------|---------------------|
| M1 | 8GB | Mini (50M) | ✅ Training + Inference |
| M1 | 16GB | Small (127M) | ✅ Training + Inference |
| M2 | 8GB | Mini (50M) | ✅ Training + Inference |
| M2 | 16GB | Small (127M) | ✅ Training + Inference |
| M2 | 24GB | Medium (350M) | ✅ Training + Inference |
| M3 | 8GB | Small (127M, quant) | ✅ Inference only |
| M3 | 16GB | Small (127M) | ✅ Training + Inference |
| M3 Max | 36GB | Large (500M+) | ✅ All operations |

## Resources

- **PyTorch MPS docs:** https://pytorch.org/docs/stable/notes/mps.html
- **Apple ML Compute:** https://developer.apple.com/documentation/metalperformanceshaders
- **Quantization guides:** bitsandbytes, GPTQ, AWQ
- **Flash Attention:** https://github.com/Dao-AILab/flash-attention

---

**Priority:** High - Critical for real-world deployment  
**Difficulty:** Medium-High - Requires low-level optimization  
**Impact:** Critical - Determines hardware accessibility  
**Timeline:** 5 weeks
